### PR TITLE
storage: handling missing keys with ENVELOPE NONE #12275

### PR DESCRIPTION
**Pull request by pH14**
This PR updates Kafka sources with `ENVELOPE NONE` to produce null key column values when the Kafka message's key is missing, rather than erroring out like it currently does. This follows the [thought noted here](https://github.com/MaterializeInc/materialize/issues/12275#issuecomment-1146726156) that it should apply equally to each message format, which is an opinion I'd be happy to revisit if there's concern.

The approach is simple -- if we see an incoming record without a key, we create a `Row` with as many `NULL` values as there are key-derived columns to fill in the blanks.

### Motivation

See #12275

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

* Added explicit tests for various permutations of Kafka messages with and without keys and values when using `ENVELOPE NONE`. This required a small tweak to `testdrive` to send messages without keys or values.
* Added tests that demonstrate successful null value handling across various formats, like Regex, Avro, Protobuf and Bytes, and when multiple key columns are flattened.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - When `INCLUDE KEY` is used with an `ENVELOPE NONE` Kafka source, records with missing keys now produce `NULL` column values, rather than raising an error.

### CODE REVIEW

File: test/testdrive/kafka-envelope-none.td
Comment (guswynn) [inline test/testdrive/kafka-envelope-none.td:59]: very thorough testing!!

File: src/testdrive/src/action/kafka/ingest.rs
Comment (guswynn) [inline src/testdrive/src/action/kafka/ingest.rs:418]:
```suggestion
                match (self.omit_key, &key_transcoder) {
                    (true, _) => None,
                    (false, None) => None,
                    (false, Some(kt)) => kt.transcode(&mut row)?
                }
```

something like this should work for the value transcoder too, but that one is simpler so your call

File: src/storage/src/render/sources.rs
Comment (guswynn) [inline src/storage/src/render/sources.rs:615]: just make this is a normal function that takes in `key_arity` and `KV`, and return a `KV`

then, above you can do:
```
.map(|kv| handle_missing_keys_fn(key_arity, kv))
```
and rustc will setup the closure that borrows `key_arity` for you!

File: src/storage/src/render/sources.rs
Comment (guswynn) [inline src/storage/src/render/sources.rs:616]: shouldn't this be happening in `flatten_results_prepend_key`? I think instead we should change the api of `raise_key_value_errors` to:
```
fn raise_key_value_errors(KV { key, val }: KV) -> Option<Result<(Option<Row>, Row, Diff), DecodeError>> {
```
and change `flatten_results_prepend_keys` to pack the key slice of `nulls`

Comment (guswynn) [reply src/storage/src/render/sources.rs:616]: additionally, we should make a `RowPacker` to avoid some extra allocations here, but that can come after my other comments!

**Review by guswynn — CHANGES_REQUESTED** [review]
Looks good! have a few code structure comments!

File: src/storage/src/render/sources.rs
Comment (pH14) [reply src/storage/src/render/sources.rs:615]: oh yeah the round-about inner function was a way to create the key slice once for a given arity, so it can be more cheaply cloned for each KV. let me play around with your idea for `raise_key_value_errors` and see how it shakes out

File: src/storage/src/render/sources.rs
Comment (pH14) [reply src/storage/src/render/sources.rs:616]: would a RowPacker be more efficient than the currently `null_key_columns.clone()`? not sure how it'd be used in this case

File: src/storage/src/render/sources.rs
Comment (guswynn) [reply src/storage/src/render/sources.rs:615]: i think its gonna be the same price if you call clone or `pack_slice`!

File: src/storage/src/render/sources.rs
Comment (pH14) [reply src/storage/src/render/sources.rs]: oh yeah the new `raise_key_value_errors` is nice to work with, that simplified the change in `flatten_results_prepend_keys` quite a bit

File: src/storage/src/render/sources.rs
Comment (pH14) [reply src/storage/src/render/sources.rs]: I created a NoneEnvelope for SourceEnvelope::None to use and pass through key arity information, let me know if it passes the sniff test!

**Comment by pH14** [comment]
I think this is about ready. @andrioni @morsapaes does this behavior seem right to you?

**Comment by morsapaes** [comment]
@pH14, thanks for picking this up! I'm +1 on the behaviour and also on backporting the fix to `v0.26.2`.

**Comment by pH14** [comment]
@sploiselle just realized I should have pinged you on this before merging 🙈

**Comment by sploiselle** [comment]
@pH14 nbd--this feature does seem like it wants to be a property of `DataEncoding`, though, rather than something danging off of nothing in the planner, like

```rust
/// A description of how each row should be decoded, from a string of bytes to a sequence of
/// Differential updates.
#[derive(Arbitrary, Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
pub enum DataEncodingInner {
    Avro(AvroEncoding),
    Protobuf(ProtobufEncoding),
    Csv(CsvEncoding),
    Regex(RegexEncoding),
    Postgres,
    Bytes,
    Text,
    RowCodec(RelationDesc),
}

pub struct DataEncoding {
  pub force_nullable_columns: bool,
  pub inner: DataEncodingInner,
}
```

and then `DataEncoding::desc` performs the work that's current `plan_create_source` to make all of the keys nulalble. This seems tractable because `get_encoding` is already looking at the envelope while determining the `DataEncoding`, but if it's not lmk.

**Comment by pH14** [comment]
👍 thank you for looking! I will take a stab at that and see how it goes

**Pull request merged**
